src/github.com/matrix-org/dendrite/roomserver/query/query.go

```diff
@@ -330,6 +330,62 @@ func (r *RoomserverQueryAPI) QueryInvitesForUser(
 	return nil
 }
 
+// QueryServerAllowedToSeeEvent implements api.RoomserverQueryAPI
+func (r *RoomserverQueryAPI) QueryServerAllowedToSeeEvent(
+	request *api.QueryServerAllowedToSeeEventRequest,
+	response *api.QueryServerAllowedToSeeEventResponse,
+) error {
+	stateEntries, err := state.LoadStateAtEvent(r.DB, request.EventID)
+	if err != nil {
+		return err
+	}
+
+	stateAtEvent, err := r.loadStateEvents(stateEntries)
+	if err != nil {
+		return err
+	}
```

Comment:
It would be nice if this could be done without loading *all* the events.
I suspect we could improve the performance here by storing numeric IDs for the membership and the domain part of the state key in the events table.

Reply:
Hmm it sounds like adding the necessary indexes is going to be a lot of effort. So I'm happy if we leave a TODO and add the index in another PR.